### ISSUE TITLE
twisted.cred.credentials.DigestCredentialFactory fails to decode responses for uri which include ","

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6609]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6609|
|Type|defect|
|Created|2013-07-01 20:23:20Z|
|Branch|https://github.com/twisted/twisted/tree/digest-with-comma-6609|

The parser is a little bit too naive.  It explodes like this if "," occurs in the URI (or any other field):

```
  File "twisted/web/_auth/wrapper.py", line 114, in _authorizedResource
    credentials = factory.decode(respString, request)
  File "twisted/web/_auth/digest.py", line 54, in decode
    request.getClientIP())
  File "twisted/cred/credentials.py", line 345, in decode
    for (k, v) in [p.split('=', 1) for p in parts]:
exceptions.ValueError: need more than 1 value to unpack
```

<details><summary>Searchable metadata</summary>

```
trac-id__6609 6609
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_digest_with_comma_6609 branches/digest-with-comma-6609
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1372710200000000 1372710200000000
changetime__1372794188000000 1372794188000000
version__None None
owner__exarkun exarkun
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [39010]) Branching to 'digest-with-comma-6609'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Ha ha, all this code is in core I guess.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

This was a duplicate of [#6445](https://github.com/twisted/twisted/issues/6445), but as noted there, I closed that one because this one has a branch in review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to @exarkun|
|-|-|

Thanks for fixing this.  It's bothered me in the past and I am sad I didn't write a patch sooner.

  1. [A couple of lines in the change are too long.](http://buildbot.twistedmatrix.com/builders/twistedchecker/builds/904/steps/run-twistedchecker/logs/new%20twistedchecker%20errors)
  2. As it happens, the only other comment I've got is that the inline regex could be pulled out to a call to `re.compile`; that would fix the line-length issue as a consequence, be a little more readable, (and also speeeeeeeed).
  3. I'm going to start mentioning this every time it happens, when reviewing tickets from core Twisted people, at least; [the Windows build for this branch](http://buildbot.twistedmatrix.com/builders/winxp32-py2.7/builds/2429/steps/select/logs/stdio) [ticket:5921 hit the ENOSPC bug].  It'd be nice to finally fix that so we get reliable builds.  (Obviously, I hope, not part of this ticket.)

Please land at your discretion.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [39015]) Merge digest-with-comma-6609 

Author: exarkun
Reviewer: glyph
Fixes: [#6609](https://github.com/twisted/twisted/issues/6609)

Fix the digest authentication response parser to handle values which include
",".
